crypto/serializer/deserializer_lib.c

```diff
@@ -0,0 +1,333 @@
+        return NULL;
+    }
+    BIO_set_fp(b, fp, BIO_NOCLOSE);
+    return b;
+}
+
+int OSSL_DESERIALIZER_from_fp(OSSL_DESERIALIZER_CTX *ctx, FILE *fp)
+{
+    BIO *b = bio_from_file(fp);
+    int ret = 0;
+
+    if (b != NULL)
+        ret = OSSL_DESERIALIZER_from_bio(ctx, b);
+
+    BIO_free(b);
+    return ret;
+}
+#endif
+
+int OSSL_DESERIALIZER_CTX_add_deserializer(OSSL_DESERIALIZER_CTX *ctx,
+                                           OSSL_DESERIALIZER *deser)
+{
+    OSSL_DESERIALIZER_INSTANCE *deser_inst = NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    OSSL_PARAM params[2];
+    void *provctx = NULL;
+
+    if (!ossl_assert(ctx != NULL) || !ossl_assert(deser != NULL)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (deser->get_params == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER,
+                  OSSL_DESERIALIZER_R_MISSING_GET_PARAMS);
+        return 0;
+    }
+
+    if (ctx->deser_insts == NULL
+        && (ctx->deser_insts =
+            sk_OSSL_DESERIALIZER_INSTANCE_new_null()) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    if ((deser_inst = OPENSSL_zalloc(sizeof(*deser_inst))) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    if (!OSSL_DESERIALIZER_up_ref(deser)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    deser_inst->deser = deser;
+
+    prov = OSSL_DESERIALIZER_provider(deser_inst->deser);
+    provctx = OSSL_PROVIDER_get0_provider_ctx(prov);
+
+    /* Cache the input type for this serializer */
+    params[0] =
+        OSSL_PARAM_construct_utf8_ptr(OSSL_DESERIALIZER_PARAM_INPUT_TYPE,
+                                      (char **)&deser_inst->input_type, 0);
+    params[1] = OSSL_PARAM_construct_end();
+
+    if (!deser_inst->deser->get_params(params)
+        || !OSSL_PARAM_modified(&params[0]))
+        goto err;
+
+    if ((deser_inst->deserctx = deser_inst->deser->newctx(provctx))
+        == NULL)
+        goto err;
+
+    if (sk_OSSL_DESERIALIZER_INSTANCE_push(ctx->deser_insts, deser_inst) <= 0)
+        goto err;
+
+    return 1;
+ err:
+    if (deser_inst != NULL) {
+        if (deser_inst->deser != NULL)
+            deser_inst->deser->freectx(deser_inst->deserctx);
+        OSSL_DESERIALIZER_free(deser_inst->deser);
+        OPENSSL_free(deser_inst);
+    }
+    return 0;
+}
+
+int OSSL_DESERIALIZER_CTX_add_extra(OSSL_DESERIALIZER_CTX *ctx,
+                                    OPENSSL_CTX *libctx, const char *propq)
+{
+    /*
+     * use the input-types of already fetched deserializers to find the next
+     * set of deserializer.  Do this until there are no more to be found.
+     * Do this no more than 10 times.
+     */
+    size_t i = 0, count = 0, depth = 0, end_i = 0;
+
+    end_i = OSSL_DESERIALIZER_CTX_num_deserializers(ctx);
```

Comment:
I find this mismatch in names between the count i.e. OSSL_DESERIALIZER_CTX_num_deserializers()
and the value that is iterated i.e k_OSSL_DESERIALIZER_INSTANCE_value() to be a very jarring.

Reply:
Sorry, I'm not following...

Reply:
OSSL_DESERIALIZER_CTX_num_deserializers()  & OSSL_DESERIALIZER_INSTANCE_value() dont look like they belong together as names..

Reply:
Ah, you'd prefer that I use `sk_OSSL_DESERIALIZER_INSTANCE_num()`...  ok, sure.

Reply:
Done